File: src/budget_api/models/category_mappings.py
"""Category mapping models — KPI roles, detailed sections, CRUD models."""

from __future__ import annotations

from pydantic import BaseModel

# KPI roles (category_roles.json) — 6 values
KPI_ROLES = {"income", "savings", "spend", "personal_spend", "investment", "exclude"}

# Detailed section mapping (detailed_section_mapping.json) — 10 values
DETAILED_CATEGORY_SECTIONS = {
    "income_salary",
    "income_third_party",
    "savings",
    "home",
    "common",
    "personal_partner_a",
    "personal_partner_b",
    "trips",
    "cc_payments",
    "excluded",
}


class CategoryMapping(BaseModel):
    """One category — id, title, kpi_role, detailed_section."""

    category_id: str
    category_title: str
    kpi_role: str | None
    detailed_section: str | None


class CategoryMappingList(BaseModel):
    """GET /api/category-mappings — merged list."""

    categories: list[CategoryMapping]


class CategoryMappingUpdate(BaseModel):
    """PUT /api/category-mappings/{id} — update both role + section."""

    kpi_role: str
    detailed_section: str

File: client/src/types/category_mappings.ts
// TS types — match backend Pydantic models (src/budget_api/models/category_mappings.py).

export type KpiRole =
  | "income"
  | "savings"
  | "spend"
  | "personal_spend"
  | "investment"
  | "exclude";

export type DetailedSection =
  | "income_salary"
  | "income_third_party"
  | "savings"
  | "home"
  | "common"
  | "personal_partner_a"
  | "personal_partner_b"
  | "trips"
  | "cc_payments"
  | "excluded";

export interface CategoryMapping {
  category_id: string;
  category_title: string;
  kpi_role: KpiRole | null;
  detailed_section: DetailedSection | null;
}

export interface CategoryMappingList {
  categories: CategoryMapping[];
}

export interface CategoryMappingUpdate {
  kpi_role: KpiRole;
  detailed_section: DetailedSection;
}

// Option lists for dropdowns — order matches design doc.
export const KPI_ROLE_OPTIONS: KpiRole[] = [
  "income",
  "savings",
  "spend",
  "personal_spend",
  "investment",
  "exclude",
];

export const DETAILED_SECTION_OPTIONS: DetailedSection[] = [
  "income_salary",
  "income_third_party",
  "savings",
  "home",
  "common",
  "personal_partner_a",
  "personal_partner_b",
  "trips",
  "cc_payments",
  "excluded",
];

// Display label for a detailed-section option. Personal sections resolve
// from config-driven partner labels (payload) — missing labels fall back
// to neutral "Partner A/B". Other sections keep their raw contract value.
export function detailedSectionLabel(
  section: DetailedSection,
  partnerLabels: Record<string, string>,
): string {
  if (section === "personal_partner_a") {
    return `Personal — ${partnerLabels["partner_a"] ?? "Partner A"}`;
  }
  if (section === "personal_partner_b") {
    return `Personal — ${partnerLabels["partner_b"] ?? "Partner B"}`;
  }
  return section;
}